apps/fz_http/lib/fz_http/rules.ex

```diff
@@ -61,72 +60,41 @@ defmodule FzHttp.Rules do
     )
   end
 
-  def nftables_device_spec(device) do
-    rules = FzHttp.Rules.list_rules(device.user_id)
-
-    Enum.map(rules, fn rule ->
-      {get_device_ip(device, ip_type("#{rule.destination}")), decode(rule.destination),
-       rule.action}
-    end)
-    # XXX: This should only be needed if a destination can be ipv4/v6 while ipv4/v6
-    # is disabled in firezone
-    |> Enum.reject(fn {source, _, _} -> is_nil(source) end)
-    |> Enum.map(fn {s, d, r} -> {decode(s), d, r} end)
+  def device_projection(device) do
+    %{ip: decode(device.ipv4), ip6: decode(device.ipv6), user_id: device.user_id}
   end
 
-  def nftables_spec(rule) do
-    nftables_spec(rule, rule.user_id)
+  def rule_projection(rule) do
+    %{destination: decode(rule.destination), user_id: rule.user_id, action: rule.action}
   end
 
-  defp nftables_spec(rule, nil) do
-    [spec_tuple(rule, nil)]
+  def user_projection(user) do
+    user.id
   end
 
-  defp nftables_spec(rule, user_id) do
-    FzHttp.Devices.list_devices(user_id)
-    |> Enum.map(fn device -> spec_tuple(rule, device) end)
-    # XXX: This should only be needed if a destination can be ipv4/v6 while ipv4/v6
-    # is disabled in firezone
-    |> Enum.reject(fn ip -> is_nil(ip) end)
+  def to_settings do
```

Comment:
Sounds good. Yeah it’s usually just run once unless the application is crashlooping, but once we have the REST API added we may see large amounts of devices / rules being added (100k+… would need to switch to eBPF) — hoping to prevent any major unnecessary db bottlenecks